File: sdk/cpp/examples/sequence_features.cc
#include <iostream>
#include <memory>
#include <vector>

#include "temporalstore/client.h"

int main(int argc, char** argv) {
    if (argc < 4) {
        std::cerr << "usage: " << argv[0] << " <metaserver_host:port> <namespace> <table>\n";
        return 2;
    }

    temporalstore::ClientOptions options;
    options.metaserver_addr = argv[1];
    options.namespace_name = argv[2];
    options.table_name = argv[3];
    options.psm = "temporalstore.cpp.sdk.example";

    std::unique_ptr<temporalstore::Client> client;
    temporalstore::Status status = temporalstore::Client::Connect(options, &client);
    if (!status.ok()) {
        std::cerr << "connect failed: " << status.ToString() << "\n";
        return 1;
    }

    const std::string key = "cpp:user:42:sequence";
    temporalstore::SequenceFeatureRow first;
    first.timestamp = 1700000000000ULL;
    first.gid = 900ULL;
    first.action_type = 1;
    first.duration = 31;
    first.author_id = 7000ULL;
    temporalstore::SequenceFeatureRow second;
    second.timestamp = 1700000001000ULL;
    second.gid = 901ULL;
    second.action_type = 3;
    second.duration = 120;
    second.author_id = 7001ULL;
    std::vector<temporalstore::SequenceFeatureRow> rows = {first, second};
    status = client->AddSequenceFeatureRows(key, rows);
    if (!status.ok()) {
        std::cerr << "add sequence failed: " << status.ToString() << "\n";
        return 1;
    }

    temporalstore::FeatureQuery query;
    query.start_ts = 1700000000000ULL;
    query.end_ts = 1700000002000ULL;
    query.count = 10;
    temporalstore::FeatureFilter filter;
    filter.field = "action_type";
    filter.op = temporalstore::FeatureFilterOp::kEqual;
    filter.value = 3;
    query.filters.push_back(filter);

    std::vector<temporalstore::SequenceFeatureRow> out;
    status = client->QuerySequenceFeatureRows(key, query, &out);
    if (!status.ok()) {
        std::cerr << "query sequence failed: " << status.ToString() << "\n";
        return 1;
    }

    std::cout << "rows=" << out.size() << "\n";
    return 0;
}

File: sdk/python/examples/sequence_features.py
from temporalstore import (
    Client,
    FeatureFilter,
    FeatureFilterOp,
    Options,
    RiskPrecision,
    SequenceFeatureRow,
    WindowUnit,
)


options = Options(
    metaserver_addr="127.0.0.1:18200",
    namespace_name="sdk_ns",
    table_name="sdk_table",
)

with Client(options) as client:
    client.put_string("python:user:42", '{"tier":"gold"}')
    print("profile=", client.get_string("python:user:42"))
    client.expire("python:user:42", 60000)
    print("ttl_ms=", client.ttl("python:user:42"))

    client.hset("python:user:42:features", "ctr_7d", "0.042")
    print("ctr_7d=", client.hget("python:user:42:features", "ctr_7d"))

    client.sadd("python:user:42:campaigns", "campaign_100")
    print("campaigns=", client.smembers("python:user:42:campaigns"))

    key = "python:user:42:sequence"
    client.add_sequence_feature_rows(
        key,
        [
            SequenceFeatureRow(1700000000000, 900, 1, 31, 7000),
            SequenceFeatureRow(1700000001000, 901, 3, 120, 7001),
        ],
    )
    rows = client.query_sequence_feature_rows(
        key,
        1700000000000,
        1700000002000,
        10,
        [FeatureFilter("action_type", FeatureFilterOp.EQUAL, 3)],
    )
    print(rows)

    risk_key = "python:user:42:risk"
    client.risk_increment(risk_key, precision=RiskPrecision.ONE_MINUTE, uuid="python-risk-1")
    print("risk_count=", client.risk_count(risk_key, window_unit=WindowUnit.HOUR))
